### PR TITLE
Implementation of batch_mimc

### PR DESCRIPTION
**Pull request by hanyunx**
The batch version of `mimc_mpc` takes a list of messages `xs = [m1, ..., mL]` and the secret key `k` as input and outputs the list `[F_MiMC(m1, k), ..., F_MiMC(mL, k)]`.

### CODE REVIEW

File: honeybadgermpc/progs/mimc.py [hunk 1]
```diff
@@ -30,9 +38,36 @@
...
+    pp_elements = PreProcessedElements()
+
+    # def cubing_share_array(): [x1,..., xK] -> [x1^3,..., xK^3]
+    async def cubing_share_array(xs):
+        rs = [[], [], []]
+        ys_, x3s = [], []
+
+        for x in xs:
+            r1, r2, r3 = pp_elements.get_cube(context)
+            y = await (x - r1).open()
```
Comment (Drake-Eidukas) [inline honeybadgermpc/progs/mimc.py:55]: I think we can do this better with a gather--
```
rs = rs_sq = cs_cube = []
for _ in range(len(xs)):
    r1, r2, r3 =pp_elements.get_cube(context)
    rs.append(r1)
    rs_sq.append(r2)
    rs_cube.append(r3)
rs = context.Share_Array(rs)
rs_sq = context.Share_Array(rs_sq)
rs_cub = context.Share_Array(rs_cub)
ys = await (await(xs - rs).open())
etc....
```

**Review by Drake-Eidukas — CHANGES_REQUESTED** [review]
One change, otherwise looks good to me.

File: honeybadgermpc/progs/mimc.py (hunk 1 above)
Comment (hanyunx) [reply honeybadgermpc/progs/mimc.py:55]: Good idea.

File: tests/progs/test_mimc.py [hunk 2]
```diff
@@ -20,13 +20,35 @@
...
+        field = GF(Subgroup.BLS12_381)
+        key = field(15)
+
+        # Compute F_MiMC_mpc, mm - mimc_mpc
+        mm = await mimc_mpc_batch(context, xs, key)
+        mm_open = await context.ShareArray(mm).open()
+
+        # open x, then compute F_MiMC_plain, mp - mimc_plain
+        xs_open = await context.ShareArray(xs).open()
+        mp = [mimc_plain(x, key) for x in xs_open]
```
Comment (amiller) [inline tests/progs/test_mimc.py:50]: Awesome test. I like how this clearly compares the batch MPC operation against the plain single version

File: tests/progs/test_mimc.py [hunk 3]
```diff
@@ -20,13 +20,35 @@
...
+    await test_runner(_prog, n, t, PREPROCESSING, k, MIXINS)
+
+
+@mark.asyncio
+async def test_mimc_mpc_batch(test_preprocessing, test_runner):
+
+    async def _prog(context):
+        xs = [test_preprocessing.elements.get_rand(context) for _ in range(20)]
+        field = GF(Subgroup.BLS12_381)
+        key = field(15)
```
Comment (amiller) [inline tests/progs/test_mimc.py]: can you pick key as a random element here rather than 15?

**Review by amiller — APPROVED** [review]
This looks really good to me

File: tests/progs/test_mimc.py (hunk 3 above)
Comment (hanyunx) [reply tests/progs/test_mimc.py]: Sure.

**Review by Drake-Eidukas — APPROVED** [review]

**Pull request merged**